src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,260 @@
+
+      pObjective(idx(j)) = function.Evaluate(pPosition.slice(idx(j)), 0);
+      for (size_t f = 1; f < numFunctions; ++f)
+        pObjective(idx(j)) += function.Evaluate(pPosition.slice(idx(j)), f);
+    }
+
+    // Sort population.
+    idx = sort_index(pObjective);
+
+    step = w(0) * pStep.slice(idx(0));
+    for (size_t j = 1; j < mu; ++j)
+      step += w(j) * pStep.slice(idx(j));
+
+    mPosition.slice(idx1) = mPosition.slice(idx0) + sigma(idx0) * step;
+
+    currentObjective = function.Evaluate(mPosition.slice(idx1), 0);
+    for (size_t f = 1; f < numFunctions; ++f)
+        currentObjective += function.Evaluate(mPosition.slice(idx1), f);
+
+    // Update best parameters.
+    if (currentObjective < overallObjective)
+    {
+      overallObjective = currentObjective;
+      iterate = mPosition.slice(idx1);
+    }
+
+    // Update Step Size.
+    if (iterate.n_rows > iterate.n_cols)
+    {
+      ps.slice(idx1) = (1 - cs) * ps.slice(idx0) + std::sqrt(
+          cs * (2 - cs) * muEffective) * covLower.t() * step;
+    }
+    else
+    {
+      ps.slice(idx1) = (1 - cs) * ps.slice(idx0) + std::sqrt(
+          cs * (2 - cs) * muEffective) * step * covLower.t();
+    }
+
+    const double psNorm = arma::norm(ps.slice(idx1));
+    sigma(idx1) = sigma(idx0) * std::pow(
+        std::exp(cs / ds * psNorm / enn - 1), 0.3);
+
+    // Update covariance matrix.
+    if ((psNorm / sqrt(1 - std::pow(1 - cs, 2 * i))) < h)
+    {
+      pc.slice(idx1) = (1 - cc) * pc.slice(idx0) + std::sqrt(cc * (2 - cc) *
+        muEffective) * step;
+
+
+      if (iterate.n_rows > iterate.n_cols)
+      {
+        C.slice(idx1) = (1 - c1 - cmu) * C.slice(idx0) + c1 *
+          (pc.slice(idx1) * pc.slice(idx1).t());
+      }
+      else
+      {
+        C.slice(idx1) = (1 - c1 - cmu) * C.slice(idx0) + c1 *
+          (pc.slice(idx1).t() * pc.slice(idx1));
+      }
+    }
+    else
+    {
+      pc.slice(idx1) = (1 - cc) * pc.slice(idx0);
+
+      if (iterate.n_rows > iterate.n_cols)
+      {
+        C.slice(idx1) = (1 - c1 - cmu) * C.slice(idx0) + c1 * (pc.slice(idx1) *
+            pc.slice(idx1).t() + (cc * (2 - cc)) * C.slice(idx0));
+      }
+      else
+      {
+        C.slice(idx1) = (1 - c1 - cmu) * C.slice(idx0) + c1 *
+            (pc.slice(idx1).t() * pc.slice(idx1) + (cc * (2 - cc)) *
+            C.slice(idx0));
+      }
+    }
+
+    if (iterate.n_rows > iterate.n_cols)
+    {
+      for (size_t j = 0; j < mu; ++j)
+      {
+        C.slice(idx1) = C.slice(idx1) + cmu * w(j) *
+            pStep.slice(idx(j)) * pStep.slice(idx(j)).t();
+      }
+    }
+    else
+    {
+      for (size_t j = 0; j < mu; ++j)
+      {
+        C.slice(idx1) = C.slice(idx1) + cmu * w(j) *
+            pStep.slice(idx(j)).t() * pStep.slice(idx(j));
+      }
+    }
+
+    arma::eig_sym(eigval, eigvec, C.slice(idx1));
+    if (arma::any(eigval < 0))
+    {
+      eigval = arma::max(eigval, eigvalZero);
+      C.slice(idx1) = eigvec * diagmat(eigval) / eigvec;
```

Comment:
Instead of `/ eigvec` should this be `* eigvec.t()`? My impression is that you are reassembling this matrix after removing all negative eigenvalues.

Another option, since `eig_sym()` gives back eigenvalues in order, is to find the first negative eigenvalue and do something like

```
C.slice(idx1) = eigvec.cols(0, firstNegativeEigenvalue - 1) * arma::diagmat(eigval.subvec(0, firstNegativeEigenvalue - 1) * eigvec.cols(0, firstNegativeEigenvalue - 1).t();
```

But if the calculation you are doing is different than what I thought, maybe my comments do not apply. :)

Reply:
Nice, catch and I like the truncated operation, I'll push the changes soon.